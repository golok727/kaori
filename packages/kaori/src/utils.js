export function invariant(condition, message) {
    if (!condition) {
        throw new Error(message);
    }
}
//# sourceMappingURL=utils.js.map