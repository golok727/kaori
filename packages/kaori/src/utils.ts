export function invariant(
  condition: boolean,
  message: string
): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}

export const has_own = Object.hasOwn;
export const get_own_property_descriptor = Object.getOwnPropertyDescriptor;
export const define_property = Object.defineProperty;
