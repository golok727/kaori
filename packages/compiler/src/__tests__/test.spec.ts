import { describe, it, expect } from 'vitest';
import { compile } from './utils';

describe('JSX to lit-html transformation', () => {
  it('should transform basic nested elements', async () => {
    const input = `
function Basic() {
  return (
    <div className="container">
      <header>
        <h1>Title</h1>
        <nav>
          <a href="/home">Home</a>
          <a href="/about">About</a>
        </nav>
      </header>
      <main>
        <p>Content here</p>
      </main>
    </div>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform mixed children (text, expressions, elements)', async () => {
    const input = `
function MixedChildren() {
  const name = "World";
  const count = 42;
  return (
    <div>
      Hello {name}!
      <span>Count: {count}</span>
      Some more text
      <p>Paragraph</p>
      Final text
    </div>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform components with children', async () => {
    const input = `
function ComponentChildren() {
  return (
    <Layout>
      <Header title="My App" />
      <Content>
        <Article title={getTitle()}>
          <p>Article content</p>
          <Comments comments={getComments()} />
        </Article>
      </Content>
      <Footer />
    </Layout>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform fragments', async () => {
    const input = `
function FragmentTest() {
  return (
    <>
      <h1>Title</h1>
      <p>Paragraph 1</p>
      <p>Paragraph 2</p>
    </>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform complex nesting with all attribute types', async () => {
    const input = `
function ComplexNesting() {
  return (
    <div className="outer">
      <input
        type="text"
        prop:value={getValue()}
        bool:disabled={isDisabled()}
        onChange={handleChange}
      />
      <Show when={showContent()}>
        <div className="inner">
          <For items={items()}>
            {(item) => (
              <Card key={item.id}>
                <h3>{item.title}</h3>
                <p>{item.description}</p>
                <button onClick={() => handleClick(item.id)}>
                  {item.completed ? "Undo" : "Complete"}
                </button>
              </Card>
            )}
          </For>
        </div>
      </Show>
    </div>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform self-closing elements', async () => {
    const input = `
function SelfClosing() {
  return (
    <div>
      <img src="/logo.png" alt="Logo" />
      <input type="text" prop:value={text()} />
      <br />
      <hr className="divider" />
      <meta name="description" content="Test" />
    </div>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform empty children and whitespace', async () => {
    const input = `
function EmptyChildren() {
  return (
    <div>
      <Empty />

      <WithChildren>

      </WithChildren>

      <p></p>
    </div>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should handle import thing', async () => {
    const input = `
function ImportThing() {
return <div><ImportThing/></div>;
}
`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should handle getter wrapping for complex expressions', async () => {
    const input = `
function GetterWrapping() {
  const count = { value: 5 };
  return (
    <div>
      <Component when={count.value > 10} />
      <Component count={count.value} />
      <Component onClick={handleClick} />
      <Component data={getData()} />
      <Component isActive={user.isLoggedIn && user.role === 'admin'} />
      <Component simple="static string" />
      <Component number={42} />
      <Component condition={a > b && c.value < d.getValue()} />
    </div>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should handle component props with colons (namespaced attributes)', async () => {
    const input = `
function NamespacedProps(props) {
  return (
    <Input
      prop:value={props.value}
      prop:checked={props.checked}
      bind:value={state.value}
    />
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
    expect(output).toContain('get ["prop:value"]()');
    expect(output).toContain('get ["prop:checked"]()');
    expect(output).toContain('get ["bind:value"]()');
  });

  it('should handle component props with hyphens (data attributes)', async () => {
    const input = `
function DataAttributes(props) {
  return (
    <Input
      data-thing="asd"
      data-value={props.value}
      aria-label="Input field"
      custom-attr={getValue()}
    />
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
    expect(output).toContain('"data-thing": "asd"');
    expect(output).toContain('get ["data-value"]()');
    expect(output).toContain('"aria-label": "Input field"');
    expect(output).toContain('get ["custom-attr"]()');
  });

  it('should handle mixed valid and invalid identifier props', async () => {
    const input = `
function MixedProps(props) {
  return (
    <Input
      thing={props.value}
      prop:value={props.value}
      data-thing="asd"
      validName={props.name}
      another-invalid={getStuff()}
    />
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
    expect(output).toContain('get thing()');
    expect(output).toContain('get ["prop:value"]()');
    expect(output).toContain('"data-thing": "asd"');
    expect(output).toContain('get validName()');
    expect(output).toContain('get ["another-invalid"]()');
  });

  it('should handle HTML elements with invalid identifier attributes', async () => {
    const input = `
function HtmlWithDataAttrs() {
  return (
    <div
      data-test-id="container"
      data-value={getValue()}
      aria-label="Main container"
      className="box"
    >
      <input data-input="text" type="text" />
    </div>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });
});
