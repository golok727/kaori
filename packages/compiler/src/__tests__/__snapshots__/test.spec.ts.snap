// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JSX to lit-html transformation > should handle getter wrapping for complex expressions 1`] = `
"import { component, html } from "kaori.js";
function GetterWrapping() {
  const count = {
    value: 5
  };
  return html\`<div>\${component(Component, {
    get when() {
      return count.value > 10;
    }
  })}\${component(Component, {
    get count() {
      return count.value;
    }
  })}\${component(Component, {
    onClick: handleClick
  })}\${component(Component, {
    get data() {
      return getData();
    }
  })}\${component(Component, {
    get isActive() {
      return user.isLoggedIn && user.role === 'admin';
    }
  })}\${component(Component, {
    simple: "static string"
  })}\${component(Component, {
    number: 42
  })}\${component(Component, {
    get condition() {
      return a > b && c.value < d.getValue();
    }
  })}</div>\`;
}"
`;

exports[`JSX to lit-html transformation > should handle import thing 1`] = `
"import { component, html } from "kaori.js";
function ImportThing() {
  return html\`<div>\${component(ImportThing, {})}</div>\`;
}"
`;

exports[`JSX to lit-html transformation > should transform basic nested elements 1`] = `
"import { html } from "kaori.js";
function Basic() {
  return html\`<div class="container"><header>
        <h1>Title</h1>
        <nav>
          <a href="/home">Home</a>
          <a href="/about">About</a>
        </nav>
      </header><main>
        <p>Content here</p>
      </main></div>\`;
}"
`;

exports[`JSX to lit-html transformation > should transform complex nesting with all attribute types 1`] = `
"import { component, html } from "kaori.js";
function ComplexNesting() {
  return html\`<div class="outer"><input type="text" .value=\${getValue()} ?disabled=\${isDisabled()} @change=\${handleChange} />\${component(Show, {
    get when() {
      return showContent();
    },
    get children() {
      return html\`<div class="inner">\${component(For, {
        get items() {
          return items();
        },
        children: item => component(Card, {
          get key() {
            return item.id;
          },
          get children() {
            return [html\`<h3>\${item.title}</h3>\`, html\`<p>\${item.description}</p>\`, html\`<button @click=\${() => handleClick(item.id)}>\${item.completed ? "Undo" : "Complete"}</button>\`];
          }
        })
      })}</div>\`;
    }
  })}</div>\`;
}"
`;

exports[`JSX to lit-html transformation > should transform components with children 1`] = `
"import { component, html } from "kaori.js";
function ComponentChildren() {
  return component(Layout, {
    get children() {
      return [component(Header, {
        title: "My App"
      }), component(Content, {
        get children() {
          return component(Article, {
            get title() {
              return getTitle();
            },
            get children() {
              return [html\`<p>Article content</p>\`, component(Comments, {
                get comments() {
                  return getComments();
                }
              })];
            }
          });
        }
      }), component(Footer, {})];
    }
  });
}"
`;

exports[`JSX to lit-html transformation > should transform empty children and whitespace 1`] = `
"import { component, html } from "kaori.js";
function EmptyChildren() {
  return html\`<div>\${component(Empty, {})}\${component(WithChildren, {})}<p></p></div>\`;
}"
`;

exports[`JSX to lit-html transformation > should transform fragments 1`] = `
"import { html } from "kaori.js";
function FragmentTest() {
  return [html\`<h1>Title</h1>\`, html\`<p>Paragraph 1</p>\`, html\`<p>Paragraph 2</p>\`];
}"
`;

exports[`JSX to lit-html transformation > should transform mixed children (text, expressions, elements) 1`] = `
"import { html } from "kaori.js";
function MixedChildren() {
  const name = "World";
  const count = 42;
  return html\`<div>
      Hello \${name}!
      <span>Count: \${count}</span>
      Some more text
      <p>Paragraph</p>
      Final text
    </div>\`;
}"
`;

exports[`JSX to lit-html transformation > should transform self-closing elements 1`] = `
"import { html } from "kaori.js";
function SelfClosing() {
  return html\`<div><img src="/logo.png" alt="Logo" /><input type="text" .value=\${text()} /><br /><hr class="divider" /><meta name="description" content="Test" /></div>\`;
}"
`;
